src/main/scala/com/microsoft/hyperspace/index/IndexLogEntry.scala

```diff
@@ -29,19 +33,172 @@ case class NoOpFingerprint() {
+    } else {
+      createSubDirectoryTree(path, Array())
+    }
+  }
+
+  /**
+   * Create a Content object from a specified list of leaf files. Any files not listed here will
+   * NOT be part of the returned object
+   * Pre-requisite: files list should be non-empty.
+   * Pre-requisite: all files must be leaf files.
+   *
+   * @param files List of leaf files.
+   * @return Content object with Directory tree from leaf files.
+   */
+  def fromLeafFiles(files: Seq[FileStatus]): Directory = {
+    require(
+      files.nonEmpty,
+      s"Empty files list found while creating a ${Directory.getClass.getName}.")
+
+    require(
+      files.forall(!_.isDirectory),
+      "All files must be leaf files for creation of Directory.")
+
+    /* from org.apache.spark.sql.execution.datasources.InMemoryFileIndex. */
+    val leafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+    val rootPath = getRoot(leafDirToChildrenFiles.head._1)
+
+    val subDirs = leafDirToChildrenFiles
+      .filterNot(_._1.isRoot)
+      .map {
+        case (dir, statuses) => createSubDirectoryTree(dir, statuses)
+      }
+      .toSeq
+
+    val rootFiles = leafDirToChildrenFiles.getOrElse(rootPath, Array()).map(FileInfo(_))
+
+    Directory(rootPath.toString, rootFiles, subDirs)
+  }
+
+  // Return file system root path from any path. E.g. "file:/C:/a/b/c" will have root "file:/C:/".
+  // For linux systems, this root will be "file:/". Other hdfs compatible file systems will have
+  // corresponding roots.
+  private def getRoot(path: Path): Path = {
+    if (path.isRoot) path else getRoot(path.getParent)
+  }
+
+  // Create leaf subdirectory. Leaf Subdirectory is the first parent of leaf files.
+  private def createSubDirectoryTree(dirPath: Path, statuses: Array[FileStatus]): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = statuses.map(FileInfo(_)))
+    } else {
+      createNonLeafSubDirectoryHelper(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = statuses.map(FileInfo(_))))
+    }
+  }
+
+  // Create Non-Leaf subdirectory tree starting from the topmost non-root directory.
+  private def createNonLeafSubDirectoryHelper(dirPath: Path, child: Directory): Directory = {
+    if (dirPath.getParent.isRoot) {
+      Directory(dirPath.getName, files = Seq(), subDirs = Seq(child))
+    } else {
+      createNonLeafSubDirectoryHelper(
+        dirPath.getParent,
+        Directory(dirPath.getName, files = Seq(), subDirs = Seq(child)))
+    }
+  }
+
+  private def listLeafFiles(
+      path: Path,
+      pathFilter: PathFilter = PathUtils.DataPathFilter,
+      throwIfNotExists: Boolean = false,
+      fs: FileSystem): Seq[FileStatus] = {
+    try {
+      val (files, directories) = fs.listStatus(path).partition(_.isFile)
+      files.filter(s => pathFilter.accept(s.getPath)) ++
+        directories.flatMap(d => listLeafFiles(d.getPath, pathFilter, throwIfNotExists, fs))
```

Comment:
we can do something like this
```
private def listLeafFiles(
      path: Path,
      pathFilter: PathFilter = PathUtils.DataPathFilter,
      throwIfNotExists: Boolean = false,
      fs: FileSystem): Seq[FileStatus] = {
    try {
      val statuses: RemoteIterator[LocatedFileStatus] = fs.listFiles(path, true)
      val filesSeq: scala.collection.mutable.ArrayBuffer[FileStatus] = ArrayBuffer()
      while (statuses.hasNext) {
        val s = statuses.next()
        if (pathFilter.accept(s.getPath)) filesSeq.append(s)
      }
      
      filesSeq
    } catch {
      case _: FileNotFoundException if !throwIfNotExists => Seq()
      case e: Throwable => throw e
    }
  }
```
but since listFiles returns `RemoteIterator` which does **not** extend scala `Iterator`, we can't use `foreach` or `for (s <-fs.liststatus(path))` for cleaner code than the above.
I looked at alternate approoaches for recursive listing and generally people follow a similar approach as the original implementation.
Please let me know if the above looks ok to explore?

Reply:
It was more on the perf than cleanness. Can you add a `TODO` to explore using `listfFiles`? Thanks for exploring!

Reply:
thanks @imback82 , added.